prov/lnx/src/lnx_ops.c

```diff
@@ -0,0 +1,1036 @@
+get_rx_entry(struct local_prov_ep *cep, struct iovec *iov, void **desc,
+	size_t count, fi_addr_t addr, uint64_t tag,
+	uint64_t ignore, void *context, uint64_t flags)
+{
+	struct lnx_rx_entry *rx_entry = NULL;
+	ofi_spin_t *bplock;
+	struct ofi_bufpool *bp;
+
+	/* if lp is NULL, then we don't know where the message is going to
+	 * come from, so allocate the rx_entry from a global pool
+	 */
+	if (!cep) {
+		bp = global_recv_bp;
+		bplock = &global_bplock;
+	} else {
+		bp = cep->lpe_recv_bp;
+		bplock = &cep->lpe_bplock;
+	}
+
+	ofi_spin_lock(bplock);
+	rx_entry = (struct lnx_rx_entry *)ofi_buf_alloc(bp);
+	ofi_spin_unlock(bplock);
+	if (rx_entry) {
+		memset(rx_entry, 0, sizeof(*rx_entry));
+		if (!cep)
+			rx_entry->rx_global = true;
+		rx_entry->rx_cep = cep;
+		lnx_init_rx_entry(rx_entry, iov, desc, count, addr, tag,
+				  ignore, context, flags);
+	}
+
+	return rx_entry;
+}
+
+static inline struct lnx_rx_entry *
+lnx_remove_first_match(struct lnx_queue *q, struct lnx_match_attr *match)
+{
+	struct lnx_rx_entry *rx_entry;
+
+	ofi_spin_lock(&q->lq_qlock);
+	rx_entry = (struct lnx_rx_entry *) dlist_remove_first_match(
+			&q->lq_queue, q->lq_match_func, match);
+	ofi_spin_unlock(&q->lq_qlock);
+
+	return rx_entry;
+}
+
+static inline void
+lnx_insert_rx_entry(struct lnx_queue *q, struct lnx_rx_entry *entry)
+{
+	ofi_spin_lock(&q->lq_qlock);
+	dlist_insert_tail((struct dlist_entry *)(&entry->rx_entry),
+			  &q->lq_queue);
+	ofi_spin_unlock(&q->lq_qlock);
+}
+
+int lnx_queue_tag(struct fi_peer_rx_entry *entry)
+{
+	struct lnx_rx_entry *rx_entry = (struct lnx_rx_entry *)entry;
+	struct lnx_peer_srq *lnx_srq = (struct lnx_peer_srq*)entry->owner_context;
+
+	FI_DBG(&lnx_prov, FI_LOG_CORE,
+		"addr = %lx tag = %lx ignore = 0 found\n",
+		entry->addr, entry->tag);
+
+	lnx_insert_rx_entry(&lnx_srq->lps_trecv.lqp_unexq, rx_entry);
+
+	return 0;
+}
+
+int lnx_get_tag(struct fid_peer_srx *srx, struct fi_peer_match_attr *match,
+		struct fi_peer_rx_entry **entry)
+{
+	struct lnx_match_attr match_attr;
+	struct lnx_peer_srq *lnx_srq;
+	struct local_prov_ep *cep;
+	struct lnx_ep *lep;
+	struct lnx_rx_entry *rx_entry;
+	fi_addr_t addr = match->addr;
+	struct lnx_srx_context *srx_ctxt;
+	uint64_t tag = match->tag;
+	int rc = 0;
+
+	/* get the endpoint */
+	cep = container_of(srx, struct local_prov_ep, lpe_srx);
+	srx_ctxt = cep->lpe_srx.ep_fid.fid.context;
+	cep = srx_ctxt->srx_cep;
+	lep = srx_ctxt->srx_lep;
+	lnx_srq = &lep->le_srq;
+
+	/* The fi_addr_t is a generic address returned by the provider. It's usually
+	 * just an index or id in their AV table. When I get it here, I could have
+	 * duplicates if multiple providers are using the same scheme to
+	 * insert in the AV table. I need to be able to identify the provider
+	 * in this function so I'm able to correctly match this message to
+	 * a possible rx entry on my receive queue. That's why we need to make
+	 * sure we use the core endpoint as part of the matching key.
+	 */
```

Comment:
maybe this isn't relevant anymore but the address provided here should be the link provided address, not the core provider address so you should be fine. this probably means you can remove the ep field in the srx (I think)